### ISSUE TITLE
Signature verification - lisbaml.ts throws ERR_FAILED_TO_VERIFY_SIGNATURE 

### ISSUE THREAD

**Issue by dsimic-coco** [tngan/samlify#389]
Hello, i encountered a weird issue. When i'm receiving response for AuthnRequest ERR_FAILED_TO_VERIFY_SIGNATURE error is thrown. I experimented around package and found out that if i comment out a single line of code, the signature verification passes. The line that i commented out was line 395 in libsaml.ts - doc.removeChild(signatureNode);

https://github.com/tngan/samlify/blob/06a8b9cb92221d03727c74ed1bc4a64b095d0b1d/src/libsaml.ts#L395

Also, i was looking at xml-crypto docs and in the section for signature verification didn't see that they use that approach with removing sig node.

Is this a bug or did i misconfigured something?

I will gladly provide more info if it's needed.

**Comment by rkkatariya** [comment]
I am running into the same issue. Commenting out the line of code is not helping me.

**Comment by tngan** [comment]
@dsimic-coco @rkkatariya Which IdP are you using ? I have confirmed a test case that if the response contains CRLF character, it will cause the failed signature verification.

**Comment by d3simic** [comment]
> I am running into the same issue. Commenting out the line of code is not helping me.

@rkkatariya FYI, I referenced ts file for clarification. Try commenting out line in dist folder if you haven't.

**Comment by d3simic** [comment]
> @dsimic-coco @rkkatariya Which IdP are you using ? I have confirmed a test case that if the response contains CRLF character, it will cause the failed signature verification.

@tngan I'm using third party IdP (croatian government - NIAS service), but i don't think that info will be helpful. From my knowledge, they usualy spin their apps on win instances with .net/java stack on backend, so your deduction might prove plausible.

**Comment by rkkatariya** [comment]
@tngan There was an issue with the certificate I was using. It is working now. I was using the Samlify as Idp and Sp.

**Comment by tngan** [comment]
@d3simic Yes, it's a known issue that the CRLF will cause the failed signature verification, from the case  that I have worked with someone in the community couple months ago, he eventually made a workaround for it by removing CRLF before the parsing process begins.

**Comment by tngan** [comment]
@rkkatariya Would you mind to share what kind of issue? It could be a hint for debugging that someone may encounter later on. :)

**Comment by dsimic-coco** [comment]
> @d3simic Yes, it's a known issue that the extra CRLF will cause the failed signature verification with using samlify, from the case that I have worked with someone in the community couple months ago, he eventually made a workaround for it by removing CRLF before the parsing process begins.

@tngan Since there are no \r\n in decoded string, I think something else might be the issue. What would you need to try to debug?

**Comment by netmiller** [comment]
Any new ideas forn that.  I have also same issue: __ERR_FAILED_TO_VERIFY_SIGNATURE__
    at /Users/esa/devel/systaylor/iris3/server/node_modules/samlify/src/libsaml.ts:401:17

Is there some workaround for this ??

**Comment by SongWeiC-inf** [comment]
hi @netmiller , have you sorted out the issue? i'm having same issue after updating my nodeJS version from 10 to 12

**Comment by smali-kazmi** [comment]
@tngan can you please guide about extra CRLF in which file? because i am stuck on this thing as well

**Comment by dsimic-coco** [comment]
@tngan Any news, ideas or plans to resolve this? If not, is it ok/safe to use forked library build with mentioned line removed?

**Comment by slephant** [comment]
Debugging this error while onboarding a new IDP led me to discover they weren't signing their responses with an [exclusive canonicalization method](https://github.com/yaronn/xml-crypto#canonicalization-and-transformation-algorithms).

When it's signed in a non-exclusive canonical form, the parent document of an XML node makes a big difference to the signature. Calling `removeChild(signatureNode)` manually strips away those effects, which causes a mismatch with the original signature. An exclusive canonical form would strip them away from both sides of the equation.

I'm not totally convinced this is a samlify bug since the [SAML 2.0 spec](http://docs.oasis-open.org/security/saml/v2.0/) admits implementations should use exclusive canonicalization with or without comments, but for what it's worth, removing the line in question doesn't seem to break validation for either method.

**Comment by mihajlon97** [comment]
@tngan hey, we ran into the same issue with CRLF and fixed it with the change in the samlify library.

instead of
`verified = verified && sig.checkSignature(doc.toString());`

use
`verified = verified && sig.checkSignature(xml);`

There is probably an issue with parsing XML and retrieving it back from the DOMParser.
XML Strings are not the same and digestValue generated by xml-crypto library  does not match the one from XML, which eventually result in error in "sig.checkSignature()" 

Here is my PR: https://github.com/tngan/samlify/pull/547
